File: backend/src/agents/ConceptAnalyzerAgent.ts
import OpenAI from 'openai';
import prisma from '../db/prisma';
import dotenv from 'dotenv';
import { CanonicalizationService } from '../services/CanonicalizationService';
import { v4 as uuidv4 } from 'uuid';

dotenv.config();

/**
 * Extracted concept structure
 */
export interface ExtractedConcept {
  name: string; // Original concept name
  normalizedName: string; // Lowercase, cleaned
  parentConcept?: string; // Parent concept (if hierarchical)
  confidence: number; // 0-1
  relevance: number; // How relevant to the bookmark
}

/**
 * Concept analysis result
 */
export interface ConceptAnalysisResult {
  concepts: ExtractedConcept[];
  hierarchy: ConceptHierarchy[];
  method: 'gpt' | 'bertopic'; // Which method was used
  processingTime: number;
  cost?: number;
}

/**
 * Concept hierarchy (parent-child relationships)
 */
export interface ConceptHierarchy {
  parent: string;
  children: string[];
}

/**
 * Concept Analyzer Agent
 *
 * Hybrid approach:
 * - Tier 1: GPT-4o-mini (primary, flexible)
 * - Future: BERTopic for batch clustering across user's bookmarks
 *
 * Extracts abstract topics and concepts from bookmark content:
 * - High-level topics (e.g., "Machine Learning", "Web Development")
 * - Subtopics (e.g., "Neural Networks" under "Machine Learning")
 * - Concept hierarchies (parent-child relationships)
 *
 * Features:
 * - Concept normalization (lowercase, dedupe)
 * - Hierarchy building (detect parent-child relationships)
 * - Confidence and relevance scoring
 * - Cost tracking
 */
export class ConceptAnalyzerAgent {
  private openai: OpenAI;

  constructor() {
    this.openai = new OpenAI({
      apiKey: process.env.OPENAI_API_KEY,
    });
  }

  /**
   * Analyze concepts from bookmark content
   *
   * @param content - Text to analyze (title + summary + key points)
   * @param embedding - Vector embedding (for future BERTopic integration)
   * @returns Analysis result with concepts and hierarchy
   */
  async analyze(
    content: string,
    embedding?: number[]
  ): Promise<ConceptAnalysisResult> {
    const startTime = Date.now();

    // For now, use GPT for all analysis
    // TODO: Add BERTopic tier for batch clustering across user's bookmarks
    const result = await this.analyzeWithGPT(content);

    const processingTime = Date.now() - startTime;

    return {
      ...result,
      processingTime,
    };
  }

  /**
   * Analyze concepts using GPT-4o-mini
   */
  private async analyzeWithGPT(content: string): Promise<{
    concepts: ExtractedConcept[];
    hierarchy: ConceptHierarchy[];
    method: 'gpt';
    cost: number;
  }> {
    const prompt = `Analyze this content and extract ALL relevant concepts and topics. Be thorough and comprehensive. Focus on:

**High-Level Topics**: Broad categories (e.g., "Machine Learning", "Web Development", "Political Science", "International Relations")
**Subtopics**: More specific concepts (e.g., "Neural Networks", "React Hooks", "Opposition Movements", "Geopolitical Strategy")
**Domain-Specific Concepts**: Field-specific ideas, theories, methodologies
**Thematic Concepts**: Key themes, issues, problems discussed
**Hierarchies**: Parent-child relationships (e.g., "Opposition Strategy" → "Political Activism" → "Political Science")

Content:
"""
${content.slice(0, 4000)}
"""

Return a JSON object with:
- concepts: Array of concepts with name, parent (if subtopic), and relevance (0-1 score)
- hierarchy: Array showing parent-child relationships

Guidelines:
- Extract AT LEAST 5-10 concepts (be comprehensive, not conservative)
- Focus on abstract topics, not concrete entities (people/companies are handled elsewhere)
- Use clear, concise concept names (2-4 words max)
- Identify hierarchies where applicable
- Include both broad themes AND specific subtopics
- Assign relevance scores based on how central the concept is to the content
- Capture the full intellectual landscape of the content

Example format:
{
  "concepts": [
    {"name": "Political Opposition", "parent": null, "relevance": 0.95},
    {"name": "Civil Resistance", "parent": "Political Opposition", "relevance": 0.85},
    {"name": "Authoritarian Regimes", "parent": null, "relevance": 0.90},
    {"name": "Democratic Transitions", "parent": "Political Opposition", "relevance": 0.75},
    {"name": "Strategic Communication", "parent": null, "relevance": 0.80}
  ],
  "hierarchy": [
    {"parent": "Political Opposition", "children": ["Civil Resistance", "Democratic Transitions"]},
    {"parent": "Authoritarian Regimes", "children": []}
  ]
}`;

    const response = await this.openai.chat.completions.create({
      model: 'gpt-4o-mini',
      messages: [
        {
          role: 'system',
          content:
            'You are an expert at identifying abstract concepts and building knowledge hierarchies. Extract only clear, meaningful concepts that help organize and understand content.',
        },
        {
          role: 'user',
          content: prompt,
        },
      ],
      response_format: { type: 'json_object' },
      temperature: 0.2, // Low temperature for consistency
      max_tokens: 800,
    });

    // Parse response
    const responseText = response.choices[0].message.content || '{"concepts": [], "hierarchy": []}';
    let rawConcepts: Array<{ name: string; parent?: string | null; relevance?: number }> = [];
    let rawHierarchy: ConceptHierarchy[] = [];

    try {
      const parsed = JSON.parse(responseText);
      rawConcepts = parsed.concepts || [];
      rawHierarchy = parsed.hierarchy || [];
    } catch (error) {
      console.error('[ConceptAnalyzer] Failed to parse GPT response:', error);
      rawConcepts = [];
      rawHierarchy = [];
    }

    // Normalize and deduplicate concepts
    const concepts = this.normalizeAndDeduplicate(
      rawConcepts.map((c) => ({
        name: c.name,
        parentConcept: c.parent || undefined,
        relevance: c.relevance || 0.7,
      }))
    );

    // Calculate cost
    const cost = this.calculateCost(
      response.usage?.prompt_tokens || 0,
      response.usage?.completion_tokens || 0
    );

    return {
      concepts,
      hierarchy: rawHierarchy,
      method: 'gpt',
      cost,
    };
  }

  /**
   * Normalize concept names and deduplicate
   */
  private normalizeAndDeduplicate(
    rawConcepts: Array<{ name: string; parentConcept?: string; relevance: number }>
  ): ExtractedConcept[] {
    const conceptMap = new Map<string, ExtractedConcept>();

    for (const concept of rawConcepts) {
      const normalized = this.normalizeConceptName(concept.name);

      if (!normalized) continue; // Skip invalid concepts

      // Normalize parent concept name too
      const normalizedParent = concept.parentConcept
        ? this.normalizeConceptName(concept.parentConcept)
        : undefined;

      if (conceptMap.has(normalized)) {
        // Merge duplicate concept (take higher relevance)
        const existing = conceptMap.get(normalized)!;
        existing.relevance = Math.max(existing.relevance, concept.relevance);
      } else {
        // Add new concept
        conceptMap.set(normalized, {
          name: concept.name,
          normalizedName: normalized,
          parentConcept: normalizedParent,
          confidence: 0.85, // GPT-4o-mini is reliable
          relevance: concept.relevance,
        });
      }
    }

    // Return concepts sorted by relevance (most relevant first)
    return Array.from(conceptMap.values()).sort((a, b) => b.relevance - a.relevance);
  }

  /**
   * Normalize concept name
   */
  private normalizeConceptName(name: string): string | null {
    let normalized = name.trim();

    // Skip empty or very short names
    if (normalized.length < 2) return null;

    // Convert to lowercase for storage (standardization)
    normalized = normalized.toLowerCase();

    // Skip overly generic terms
    const genericTerms = [
      'topic',
      'concept',
      'idea',
      'thing',
      'stuff',
      'content',
      'information',
      'data',
    ];
    if (genericTerms.includes(normalized)) return null;

    // Clean whitespace
    normalized = normalized.replace(/\s+/g, ' ').trim();

    return normalized;
  }

  /**
   * Calculate GPT cost
   * GPT-4o-mini pricing: $0.150 per 1M input tokens, $0.600 per 1M output tokens
   */
  private calculateCost(inputTokens: number, outputTokens: number): number {
    const inputCost = (inputTokens / 1_000_000) * 0.15;
    const outputCost = (outputTokens / 1_000_000) * 0.6;
    return inputCost + outputCost;
  }

  /**
   * Save concepts to database (with hierarchy)
   *
   * @param concepts - Extracted concepts
   * @param bookmarkId - Bookmark to associate concepts with
   * @param userId - User ID for isolation
   */
  async saveConcepts(
    concepts: ExtractedConcept[],
    bookmarkId: string,
    userId: string,
    embedding?: number[]
  ): Promise<void> {
    if (concepts.length === 0) {
      console.log(`[ConceptAnalyzer] No concepts to save for bookmark ${bookmarkId}`);
      return;
    }

    console.log(`[ConceptAnalyzer] Saving ${concepts.length} concepts for bookmark ${bookmarkId}`);

    // DEDUPLICATION: Remove concepts that overlap with entities
    // Entities are more specific (e.g., "React" as technology) than concepts (e.g., "React" as topic)
    // So we prioritize entities and skip overlapping concepts to avoid redundancy
    const entityRelationships = await prisma.relationship.findMany({
      where: {
        userId,
        sourceType: 'bookmark',
        sourceId: bookmarkId,
        targetType: 'entity',
      },
    });

    if (entityRelationships.length > 0) {
      const entities = await prisma.entity.findMany({
        where: {
          id: { in: entityRelationships.map(rel => rel.targetId) },
        },
      });

      // Create set of entity names (normalized lowercase) for fast lookup
      const entityNames = new Set(
        entities.flatMap(e => [
          e.name.toLowerCase(),
          e.canonicalName.toLowerCase(),
          ...e.aliases.map(a => a.toLowerCase()),
        ])
      );

      // Filter out concepts that match entity names
      const beforeCount = concepts.length;
      concepts = concepts.filter(concept => {
        const conceptNormalized = concept.name.toLowerCase();
        const isOverlap = entityNames.has(conceptNormalized);

        if (isOverlap) {
          console.log(
            `[ConceptAnalyzer] ⚠️  Skipping concept "${concept.name}" - overlaps with entity`
          );
        }

        return !isOverlap;
      });

      const removedCount = beforeCount - concepts.length;
      if (removedCount > 0) {
        console.log(
          `[ConceptAnalyzer] ✓ Removed ${removedCount} overlapping concept(s), ${concepts.length} remain`
        );
      }
    }

    // If all concepts were filtered out, nothing left to save
    if (concepts.length === 0) {
      console.log(`[ConceptAnalyzer] No concepts to save after deduplication for bookmark ${bookmarkId}`);
      return;
    }

    const canonicalService = new CanonicalizationService();

    // Two-pass approach:
    // Pass 1: Create all concepts (without parent references)
    // Pass 2: Update parent references

    const createdConcepts = new Map<string, string>(); // canonicalName -> id
    const conceptMap = new Map<string, any>(); // Store canonicalized concepts for parent mapping

    // Pass 1: Create/update all concepts with canonicalization
    for (const concept of concepts) {
      try {
        // Phase 2: Canonicalize concept before saving
        const canonical = await canonicalService.resolveConcept(
          concept.name,
          '',  // Description will be added in Phase 3
          embedding || null,
          userId
        );

        // Add current mention to aliases
        const aliases = new Set([...canonical.aliases, concept.name]);
        if (concept.normalizedName) aliases.add(concept.normalizedName);

        const dbConcept = await prisma.concept.upsert({
          where: {
            userId_canonicalName: {
              userId,
              canonicalName: canonical.canonicalName,
            },
          },
          create: {
            id: canonical.id || uuidv4(),
            userId,
            name: canonical.canonicalName, // Use canonical name
            normalizedName: canonical.canonicalName.toLowerCase(),
            canonicalName: canonical.canonicalName,
            aliases: Array.from(aliases),
            description: canonical.description,
            popularity: 1,
            occurrenceCount: 1,
          },
          update: {
            occurrenceCount: { increment: 1 },
            popularity: { increment: 1 },
            aliases: Array.from(aliases), // Merge aliases
            description: canonical.description || undefined,
          },
        });

        console.log(
          `[ConceptAnalyzer] ✓ Saved concept "${dbConcept.canonicalName}" (ID: ${dbConcept.id})`
        );

        createdConcepts.set(canonical.canonicalName, dbConcept.id);
        conceptMap.set(concept.name, { canonical, dbConcept });

        // Create relationship: bookmark -> concept
        console.log(
          `[ConceptAnalyzer] Creating relationship: bookmark ${bookmarkId} -> concept ${dbConcept.id}`
        );
        const relationship = await prisma.relationship.upsert({
          where: {
            userId_sourceType_sourceId_targetType_targetId_relationshipType: {
              userId,
              sourceType: 'bookmark',
              sourceId: bookmarkId,
              targetType: 'concept',
              targetId: dbConcept.id,
              relationshipType: 'about',
            },
          },
          create: {
            userId,
            sourceType: 'bookmark',
            sourceId: bookmarkId,
            targetType: 'concept',
            targetId: dbConcept.id,
            relationshipType: 'about',
            weight: concept.relevance,
            metadata: {
              confidence: concept.confidence,
              canonicalMethod: canonical.method,
            },
          },
          update: {
            weight: concept.relevance,
            metadata: {
              confidence: concept.confidence,
            },
          },
        });

        console.log(
          `[ConceptAnalyzer] ✓ Created relationship: bookmark ${bookmarkId} -> concept ${dbConcept.id} (relationship ID: ${relationship.id})`
        );
      } catch (error) {
        console.error(
          `[ConceptAnalyzer] Failed to save concept ${concept.name}:`,
          error
        );
        // Continue with other concepts even if one fails
      }
    }

    // Pass 2: Update parent-child relationships
    for (const concept of concepts) {
      if (!concept.parentConcept) continue;

      try {
        // Get canonical child concept
        const childData = conceptMap.get(concept.name);
        if (!childData) continue;

        const childId = childData.dbConcept.id;

        // Canonicalize parent concept name to find its ID
        const parentCanonical = await canonicalService.resolveConcept(
          concept.parentConcept,
          '',
          embedding || null,
          userId
        );

        const parentId = createdConcepts.get(parentCanonical.canonicalName);

        if (!childId || !parentId) {
          console.warn(
            `[ConceptAnalyzer] Cannot create hierarchy: ${concept.name} -> ${concept.parentConcept} (missing IDs)`
          );
          continue;
        }

        // Update concept with parent reference
        await prisma.concept.update({
          where: { id: childId },
          data: {
            parentConceptId: parentId,
          },
        });

        // Create relationship: child concept -> parent concept
        await prisma.relationship.upsert({
          where: {
            userId_sourceType_sourceId_targetType_targetId_relationshipType: {
              userId,
              sourceType: 'concept',
              sourceId: childId,
              targetType: 'concept',
              targetId: parentId,
              relationshipType: 'related_to',
            },
          },
          create: {
            userId,
            sourceType: 'concept',
            sourceId: childId,
            targetType: 'concept',
            targetId: parentId,
            relationshipType: 'related_to',
            weight: 1.0, // Strong relationship (parent-child)
            metadata: {
              hierarchyType: 'parent-child',
            },
          },
          update: {
            weight: 1.0,
          },
        });
      } catch (error) {
        console.error(
          `[ConceptAnalyzer] Failed to create hierarchy for ${concept.name}:`,
          error
        );
        // Continue with other concepts even if one fails
      }
    }

    // Pass 3: Generate embeddings for new concepts
    // Generate embeddings for concepts that don't have them yet
    for (const concept of concepts) {
      const canonical = conceptMap.get(concept.name)?.canonical;
      if (!canonical) continue;

      const conceptId = createdConcepts.get(canonical.canonicalName);
      if (!conceptId) continue;

      const dbConcept = await prisma.concept.findUnique({
        where: { id: conceptId },
        select: { embedding: true },
      });

      if (!dbConcept?.embedding) {
        try {
          const embedding = await this.generateConceptEmbedding(conceptId, userId);
          if (embedding) {
            const embeddingStr = `[${embedding.join(',')}]`;
            await prisma.$executeRaw`
              UPDATE concepts SET embedding = ${embeddingStr}::vector
              WHERE id = ${conceptId}
            `;
            console.log(`[ConceptAnalyzer] ✓ Generated embedding for concept "${canonical.canonicalName}"`);
          }
        } catch (error) {
          console.error(`[ConceptAnalyzer] Failed to generate concept embedding:`, error);
          // Continue with other concepts even if embedding fails
        }
      }
    }

    // Cleanup canonicalization service
    await canonicalService.disconnect();

    console.log(`[ConceptAnalyzer] ✓ Concepts saved successfully with hierarchy and embeddings`);
  }

  /**
   * Find related concepts (co-occurrence analysis)
   *
   * Given a concept, find other concepts that frequently appear together
   * across the user's bookmarks.
   *
   * @param conceptId - Concept to find relations for
   * @param userId - User ID for isolation
   * @param minCoOccurrence - Minimum number of shared bookmarks (default: 2)
   */
  async findRelatedConcepts(
    conceptId: string,
    userId: string,
    minCoOccurrence: number = 2
  ): Promise<Array<{ concept: any; coOccurrenceCount: number; weight: number }>> {
    // Find bookmarks that have this concept
    const bookmarksWithConcept = await prisma.relationship.findMany({
      where: {
        userId,
        targetType: 'concept',
        targetId: conceptId,
        relationshipType: 'about',
      },
      select: {
        sourceId: true, // bookmark IDs
      },
    });

    const bookmarkIds = bookmarksWithConcept.map((r) => r.sourceId);

    if (bookmarkIds.length === 0) {
      return [];
    }

    // Find other concepts that appear in these bookmarks
    const relatedConceptRelationships = await prisma.relationship.findMany({
      where: {
        userId,
        sourceType: 'bookmark',
        sourceId: { in: bookmarkIds },
        targetType: 'concept',
        relationshipType: 'about',
        NOT: {
          targetId: conceptId, // Exclude the original concept
        },
      },
      select: {
        targetId: true,
        weight: true,
      },
    });

    // Count co-occurrences
    const coOccurrenceMap = new Map<
      string,
      { count: number; totalWeight: number }
    >();

    for (const rel of relatedConceptRelationships) {
      const existing = coOccurrenceMap.get(rel.targetId) || {
        count: 0,
        totalWeight: 0,
      };
      existing.count += 1;
      existing.totalWeight += rel.weight;
      coOccurrenceMap.set(rel.targetId, existing);
    }

    // Filter by minimum co-occurrence and fetch concept details
    const relatedConceptIds = Array.from(coOccurrenceMap.entries())
      .filter(([_, data]) => data.count >= minCoOccurrence)
      .sort((a, b) => b[1].count - a[1].count) // Sort by co-occurrence count
      .slice(0, 20); // Limit to top 20

    if (relatedConceptIds.length === 0) {
      return [];
    }

    const concepts = await prisma.concept.findMany({
      where: {
        id: { in: relatedConceptIds.map(([id]) => id) },
      },
    });

    // Merge with co-occurrence data
    return relatedConceptIds.map(([id, data]) => {
      const concept = concepts.find((c) => c.id === id);
      return {
        concept,
        coOccurrenceCount: data.count,
        weight: data.totalWeight / data.count, // Average weight
      };
    });
  }

  /**
   * Generate embedding for a concept based on its definition
   * Approach: Find top 3 bookmarks, extract summaries, embed definition
   */
  async generateConceptEmbedding(
    conceptId: string,
    userId: string
  ): Promise<number[] | null> {
    try {
      // 1. Find top 3 bookmarks about this concept
      const relationships = await prisma.relationship.findMany({
        where: {
          userId,
          targetType: 'concept',
          targetId: conceptId,
          relationshipType: 'about',
        },
        orderBy: { weight: 'desc' },
        take: 3,
      });

      if (relationships.length === 0) {
        console.warn(`Cannot generate embedding for concept ${conceptId} - no bookmarks`);
        return null;
      }

      // 2. Fetch bookmark summaries
      const bookmarkIds = relationships.map(r => r.sourceId);
      const bookmarks = await prisma.bookmark.findMany({
        where: { id: { in: bookmarkIds } },
        select: { title: true, summary: true },
      });

      // 3. Fetch concept details
      const concept = await prisma.concept.findUnique({
        where: { id: conceptId },
      });

      if (!concept) return null;

      // 4. Build context-rich definition
      const definition = bookmarks
        .map(b => b.summary || b.title)
        .filter(Boolean)
        .join('\n\n');

      // 5. Create embedding text
      const embeddingText = `Concept: ${concept.canonicalName}\n\nContext: ${definition}`;

      // 6. Generate embedding
      const { getEmbedderAgent } = await import('./embedderAgent');
      const embedder = getEmbedderAgent();
      const embedding = await embedder.embed({ text: embeddingText, useCache: true });

      console.log(`Generated embedding for concept "${concept.canonicalName}"`);
      return embedding;
    } catch (error) {
      console.error(`Failed to generate concept embedding:`, error);
      return null;
    }
  }

  /**
   * Batch generate embeddings for all concepts without embeddings
   */
  async batchGenerateConceptEmbeddings(
    userId: string,
    batchSize: number = 10
  ): Promise<{ success: number; failed: number }> {
    console.log(`Starting batch concept embedding for user ${userId}`);

    // Use raw SQL to filter by vector field (Prisma can't filter Unsupported types)
    const concepts = await prisma.$queryRaw<Array<{
      id: string;
      canonical_name: string;
    }>>`
      SELECT id, canonical_name
      FROM concepts
      WHERE user_id = ${userId} AND embedding IS NULL
    `;

    console.log(`Found ${concepts.length} concepts without embeddings`);

    let success = 0;
    let failed = 0;

    // Process in batches
    for (let i = 0; i < concepts.length; i += batchSize) {
      const batch = concepts.slice(i, i + batchSize);

      const results = await Promise.allSettled(
        batch.map(async (concept) => {
          const embedding = await this.generateConceptEmbedding(concept.id, userId);

          if (embedding) {
            const embeddingStr = `[${embedding.join(',')}]`;
            await prisma.$executeRaw`
              UPDATE concepts
              SET embedding = ${embeddingStr}::vector
              WHERE id = ${concept.id}
            `;
            return { success: true };
          }
          return { success: false };
        })
      );

      results.forEach(r => {
        if (r.status === 'fulfilled' && r.value.success) {
          success++;
        } else {
          failed++;
        }
      });

      // Rate limiting
      if (i + batchSize < concepts.length) {
        await new Promise(resolve => setTimeout(resolve, 1000));
      }
    }

    console.log(`Batch complete: ${success} success, ${failed} failed`);
    return { success, failed };
  }
}

File: ai-service/agents/concept_analyzer_agent.py
"""
Concept Analyzer Agent - Phase 2 with Canonicalization.

Analyzes abstract topics and concepts from bookmark content using GPT-4o-mini.
Implements Phase 2 improvements:
- Temperature 0.1 (was 0.2) - near-deterministic
- Post-processing normalization layer
- Standard hierarchy enforcement
- Prepares for semantic canonicalization (Phase 2.1)

Extracts high-level topics (e.g., "Machine Learning", "Web Development")
and builds concept hierarchies (parent-child relationships).
"""
import json
import logging
from typing import List, Dict, Optional, Any, Tuple
from datetime import datetime

from langchain_openai import ChatOpenAI
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from config import settings
from models import Concept, Relationship
from utils.concept_normalizer import (
    normalize_concept_name,
    fix_concept_hierarchy,
    deduplicate_concepts
)

logger = logging.getLogger(__name__)


class ExtractedConcept:
    """Extracted concept structure."""

    def __init__(
        self,
        name: str,
        normalized_name: str,
        parent_concept: Optional[str],
        confidence: float,
        relevance: float
    ):
        self.name = name
        self.normalized_name = normalized_name
        self.parent_concept = parent_concept
        self.confidence = confidence
        self.relevance = relevance


class ConceptHierarchy:
    """Concept hierarchy (parent-child relationships)."""

    def __init__(self, parent: str, children: List[str]):
        self.parent = parent
        self.children = children


class ConceptAnalysisResult:
    """Concept analysis result."""

    def __init__(
        self,
        concepts: List[ExtractedConcept],
        hierarchy: List[ConceptHierarchy],
        method: str,
        processing_time: int,
        cost: Optional[float] = None
    ):
        self.concepts = concepts
        self.hierarchy = hierarchy
        self.method = method
        self.processing_time = processing_time
        self.cost = cost


class ConceptAnalyzerAgent:
    """
    Concept Analyzer Agent with Phase 2 improvements.

    Phase 2 changes:
    - Temperature: 0.1 (was 0.2) - near-deterministic
    - Post-processing normalization using concept_normalizer
    - Standard hierarchy enforcement
    - Ready for semantic canonicalization (Phase 2.1)
    """

    def __init__(self):
        """Initialize concept analyzer with GPT-4o-mini."""
        self.model_name = settings.AI_MODEL
        self.temperature = 0.1  # Phase 2: Near-deterministic (was 0.2)

    async def analyze(
        self,
        content: str,
        embedding: Optional[List[float]] = None
    ) -> ConceptAnalysisResult:
        """
        Analyze concepts from bookmark content.

        Args:
            content: Text to analyze (title + summary + key points)
            embedding: Vector embedding (for future semantic canonicalization)

        Returns:
            ConceptAnalysisResult with extracted concepts and hierarchy
        """
        import time
        start_time = int(time.time() * 1000)

        # Use GPT for analysis
        result = await self._analyze_with_gpt(content)

        processing_time = int(time.time() * 1000) - start_time

        return ConceptAnalysisResult(
            concepts=result['concepts'],
            hierarchy=result['hierarchy'],
            method='gpt',
            processing_time=processing_time,
            cost=result.get('cost')
        )

    async def _analyze_with_gpt(self, content: str) -> Dict[str, Any]:
        """
        Analyze concepts using GPT-4o-mini with Phase 2 improvements.

        Phase 2:
        - Temperature 0.1 (near-deterministic, was 0.2)
        - JSON mode for reliable parsing
        - Post-processing normalization and hierarchy enforcement
        """
        # Truncate content to avoid excessive tokens
        truncated_content = content[:4000]

        prompt = f"""Analyze this content and extract key concepts and topics. Focus on:

**High-Level Topics**: Broad categories (e.g., "Machine Learning", "Web Development", "Psychology")
**Subtopics**: More specific concepts (e.g., "Neural Networks", "React Hooks", "Cognitive Load")
**Hierarchies**: Parent-child relationships (e.g., "React Hooks" is a subtopic of "Web Development")

Content:
\"\"\"
{truncated_content}
\"\"\"

Return a JSON object with:
- concepts: Array of concepts with name, parent (if subtopic), and relevance (0-1 score)
- hierarchy: Array showing parent-child relationships

Guidelines:
- Extract 3-8 concepts (don't over-extract)
- Focus on abstract topics, not concrete entities (people/companies are handled elsewhere)
- Use clear, concise concept names (2-4 words max)
- Identify hierarchies where applicable (e.g., "Deep Learning" → "Machine Learning" → "Artificial Intelligence")
- Assign relevance scores based on how central the concept is to the content

Example format:
{{
  "concepts": [
    {{"name": "Machine Learning", "parent": null, "relevance": 0.95}},
    {{"name": "Neural Networks", "parent": "Machine Learning", "relevance": 0.85}},
    {{"name": "Deep Learning", "parent": "Neural Networks", "relevance": 0.80}}
  ],
  "hierarchy": [
    {{"parent": "Machine Learning", "children": ["Neural Networks"]}},
    {{"parent": "Neural Networks", "children": ["Deep Learning"]}}
  ]
}}"""

        # Initialize LLM with JSON mode (Phase 2: temp 0.1)
        llm = ChatOpenAI(
            model=self.model_name,
            temperature=self.temperature,  # 0.1 for Phase 2
            max_tokens=800,
            max_retries=settings.LLM_MAX_RETRIES,
            model_kwargs={"response_format": {"type": "json_object"}}
        )

        messages = [
            {
                "role": "system",
                "content": "You are an expert at identifying abstract concepts and building knowledge hierarchies. Extract only clear, meaningful concepts that help organize and understand content."
            },
            {
                "role": "user",
                "content": prompt
            }
        ]

        try:
            response = await llm.ainvoke(messages)

            # Parse JSON response
            content_str = response.content if hasattr(response, 'content') else str(response)
            parsed = json.loads(content_str)

            raw_concepts = parsed.get('concepts', [])
            raw_hierarchy = parsed.get('hierarchy', [])

            # Normalize and deduplicate concepts (Phase 2)
            concepts = self._normalize_and_deduplicate(raw_concepts)

            # Build hierarchy objects
            hierarchy = [
                ConceptHierarchy(
                    parent=h.get('parent', ''),
                    children=h.get('children', [])
                )
                for h in raw_hierarchy
            ]

            # Calculate cost
            usage = response.response_metadata.get('token_usage', {})
            input_tokens = usage.get('prompt_tokens', 0)
            output_tokens = usage.get('completion_tokens', 0)
            cost = self._calculate_cost(input_tokens, output_tokens)

            logger.info(f"[ConceptAnalyzer] Extracted {len(concepts)} concepts (cost: ${cost:.6f})")

            return {
                'concepts': concepts,
                'hierarchy': hierarchy,
                'cost': cost
            }

        except json.JSONDecodeError as e:
            logger.error(f"[ConceptAnalyzer] Failed to parse GPT response: {e}")
            return {'concepts': [], 'hierarchy': [], 'cost': 0.0}
        except Exception as e:
            logger.error(f"[ConceptAnalyzer] Error analyzing concepts: {e}")
            return {'concepts': [], 'hierarchy': [], 'cost': 0.0}

    def _normalize_and_deduplicate(
        self,
        raw_concepts: List[Dict]
    ) -> List[ExtractedConcept]:
        """
        Normalize concept names and deduplicate.

        Phase 2: Uses concept_normalizer for post-processing.
        """
        concept_list = []

        for concept_data in raw_concepts:
            name = concept_data.get('name', '').strip()
            parent = concept_data.get('parent')
            relevance = concept_data.get('relevance', 0.7)

            # Phase 2: Post-processing normalization
            normalized = normalize_concept_name(name)

            if not normalized:
                continue  # Filtered out by normalizer

            # Normalize parent concept too
            normalized_parent = None
            if parent:
                normalized_parent = normalize_concept_name(parent)

            # Phase 2: Fix hierarchy using standard taxonomy
            normalized, normalized_parent = fix_concept_hierarchy(
                normalized,
                normalized_parent
            )

            concept_list.append({
                'name': name,
                'normalizedName': normalized,
                'parentConcept': normalized_parent,
                'relevance': relevance,
                'confidence': 0.85  # GPT-4o-mini is reliable
            })

        # Phase 2: Deduplicate using normalizer utility
        deduplicated = deduplicate_concepts(concept_list)

        # Convert to ExtractedConcept objects
        return [
            ExtractedConcept(
                name=c['name'],
                normalized_name=c['normalizedName'],
                parent_concept=c.get('parentConcept'),
                confidence=c['confidence'],
                relevance=c['relevance']
            )
            for c in deduplicated
        ]

    def _calculate_cost(self, input_tokens: int, output_tokens: int) -> float:
        """
        Calculate GPT cost.
        GPT-4o-mini pricing: $0.150 per 1M input tokens, $0.600 per 1M output tokens
        """
        input_cost = (input_tokens / 1_000_000) * 0.15
        output_cost = (output_tokens / 1_000_000) * 0.6
        return input_cost + output_cost

    async def save_concepts(
        self,
        concepts: List[ExtractedConcept],
        bookmark_id: str,
        user_id: str,
        session: AsyncSession
    ) -> None:
        """
        Save concepts to database with hierarchy.

        Two-pass approach:
        1. Create/update all concepts (without parent references)
        2. Update parent references

        Args:
            concepts: Extracted concepts
            bookmark_id: Bookmark to associate concepts with
            user_id: User ID for isolation
            session: Database session
        """
        if not concepts:
            logger.info(f"[ConceptAnalyzer] No concepts to save for bookmark {bookmark_id}")
            return

        logger.info(f"[ConceptAnalyzer] Saving {len(concepts)} concepts for bookmark {bookmark_id}")

        # Map: normalized_name → id
        created_concepts = {}

        # Pass 1: Create/update all concepts
        for concept in concepts:
            try:
                # Check if concept exists
                result = await session.execute(
                    select(Concept).where(
                        Concept.user_id == user_id,
                        Concept.normalized_name == concept.normalized_name
                    )
                )
                db_concept = result.scalar_one_or_none()

                if db_concept:
                    # Update existing concept
                    db_concept.occurrence_count += 1
                    db_concept.name = concept.name  # Keep most recent name
                else:
                    # Create new concept
                    db_concept = Concept(
                        user_id=user_id,
                        name=concept.name,
                        normalized_name=concept.normalized_name,
                        occurrence_count=1,
                        created_at=datetime.now()
                    )
                    session.add(db_concept)
                    await session.flush()  # Get concept ID

                created_concepts[concept.normalized_name] = db_concept.id

                # Create/update relationship: bookmark → concept
                rel_result = await session.execute(
                    select(Relationship).where(
                        Relationship.user_id == user_id,
                        Relationship.source_type == 'bookmark',
                        Relationship.source_id == bookmark_id,
                        Relationship.target_type == 'concept',
                        Relationship.target_id == db_concept.id,
                        Relationship.relationship_type == 'about'
                    )
                )
                db_relationship = rel_result.scalar_one_or_none()

                if db_relationship:
                    # Update existing relationship
                    db_relationship.weight = concept.relevance
                    db_relationship.relationship_metadata = {'confidence': concept.confidence}
                else:
                    # Create new relationship
                    db_relationship = Relationship(
                        user_id=user_id,
                        source_type='bookmark',
                        source_id=bookmark_id,
                        target_type='concept',
                        target_id=db_concept.id,
                        relationship_type='about',
                        weight=concept.relevance,
                        relationship_metadata={'confidence': concept.confidence}
                    )
                    session.add(db_relationship)

            except Exception as e:
                logger.error(f"[ConceptAnalyzer] Failed to save concept {concept.name}: {e}")
                # Continue with other concepts even if one fails

        # Pass 2: Update parent-child relationships
        for concept in concepts:
            if not concept.parent_concept:
                continue

            try:
                child_id = created_concepts.get(concept.normalized_name)
                parent_id = created_concepts.get(concept.parent_concept)

                if not child_id or not parent_id:
                    logger.warning(
                        f"[ConceptAnalyzer] Cannot create hierarchy: "
                        f"{concept.name} -> {concept.parent_concept} (missing IDs)"
                    )
                    continue

                # Update concept with parent reference
                result = await session.execute(
                    select(Concept).where(Concept.id == child_id)
                )
                db_concept = result.scalar_one_or_none()

                if db_concept:
                    db_concept.parent_concept_id = parent_id

                # Create/update relationship: child concept → parent concept
                rel_result = await session.execute(
                    select(Relationship).where(
                        Relationship.user_id == user_id,
                        Relationship.source_type == 'concept',
                        Relationship.source_id == child_id,
                        Relationship.target_type == 'concept',
                        Relationship.target_id == parent_id,
                        Relationship.relationship_type == 'related_to'
                    )
                )
                db_relationship = rel_result.scalar_one_or_none()

                if db_relationship:
                    # Update existing relationship
                    db_relationship.weight = 1.0
                else:
                    # Create new relationship
                    db_relationship = Relationship(
                        user_id=user_id,
                        source_type='concept',
                        source_id=child_id,
                        target_type='concept',
                        target_id=parent_id,
                        relationship_type='related_to',
                        weight=1.0,  # Strong relationship (parent-child)
                        relationship_metadata={'hierarchy_type': 'parent-child'}
                    )
                    session.add(db_relationship)

            except Exception as e:
                logger.error(
                    f"[ConceptAnalyzer] Failed to create hierarchy for {concept.name}: {e}"
                )
                # Continue with other concepts even if one fails

        await session.commit()
        logger.info(f"[ConceptAnalyzer] ✓ Concepts saved successfully with hierarchy")
